### ISSUE TITLE
[Pioneer] Missing ringtone for VoIP application when trying to play ringtone on BT headset

### ISSUE THREAD

**Issue by dimitar10** [sonyxperiadev/bug_tracker#176]
**Platform:** Nile
**Device:** Pioneer
**Kernel version:** 4.4.156-gf238a0
**Android version:** 8.1.0-r46
**Firmware version:** v15

**Description and symptoms**

When the device is connected to a BT headset, there's no ringtone playback for incoming VoIP calls, neither on speaker, nor in the BT earphone. On the other hand, using the same VoIP application there's ringtone when the XA2 device is running stock Sony firmware.

**How to reproduce**

Connect a Bluetooth (handsfree) headset, and make a VoIP call.

A limiting factor is that none of the common apps (e.g. Whatsapp, Skype, Wire etc.) try to play a ringtone via the BT headset. Instead, they are using the speakerphone or rerouting of the MediaPlayer (Viber) or vibrations to alert the user.

**Discussion**:

From the logcat and dmesg one can observe the following:

logcat:
...
E/audio_hw_primary( 641): disable_snd_device: Invalid sound device 0
...
E/audio_hw_primary( 641): start_output_stream: pcm_prepare returned -1
...

dmesg:

[ 1102.116300] q6asm_callback: cmd = 0x10bcd returned error = 0x1
[ 1102.121158] __q6asm_cmd: DSP returned error[ADSP_EFAILED] opcode 68557
[ 1102.144428] SDM660 LowLatency: ASoC: no backend DAIs enabled for SDM660 LowLatency

Because the VoIP application works without an issue on a device using stock Sony firmware, I don't believe it's misusing the Android APIs.

To sum up, although the issue does not have a big impact, perhaps it points to an issue that needs to be addressed, so possibly the logcat and dmesg will be of some value.

**Logs**: 

[bt-headset-issue-pioneer-20181001.zip](https://github.com/sonyxperiadev/bug_tracker/files/2434454/bt-headset-issue-pioneer-20181001.zip)

**Comment by dimitar10** [comment]
@richardtullberg  Linphone and Wire are also not working properly. Linphone should play ringtones on both speaker and BT headphone, but only manages to play the ringtone on the speaker (for example on a device with stock firmware it plays it on both. Please note that if you try for example with Linphone that the errors are different.

I will try to replicate the issue with a minimal .apk  that uses the same BT / Ringtone / CallManager APIs and paste it here as soon as I find some time.

**Comment by richardtullberg** [comment]
@dimitar10 you don't need to dig anymore. I can reproduce it with Linphone. I started checking the logs and saw that its only being routed to the "speaker" when using a bt-headset. I also noticed that it routed to both speaker and headphones if I had one of those connected. I checked in
hardware/qcom/audio/msm8916/platform.c
and there I saw that there is no "speake/BT" dual device audio routing like there is on the msm8974 platform.

I will continue to check why that is, maybe we just need to add that routing and it works, I will see

**Comment by richardtullberg** [comment]
Hi @dimitar10,
we have both bad and good news. The bad news is that there seems to be to many patches that needs to be done in the Audio Policy Manager to "allow" dual speaker bt-sco on android 8.1.

The Good news is that on android 9.0 all of these patches exist but on the platform side for nile we need to patch. But this is just one patch that can be cherry-picked from another platform as in we take the speaker-and-bt-sco patch from msm8974 and patch it to msm8916 and then it works fine. If you deem it necessary then the patch is this that needs to be brought to the msm8916 on android9.0

In hardware/qcom/audio

commit 6dcb1a8afb533aea509fd7ffbe7a442c9e15fdb7
Author: Haynes Mathew George <hgeorge@codeaurora.org>
Date:   Wed Dec 21 12:38:55 2016 -0800

    BT sco and speaker combo device selection logic
    
    Bug: 19171297
    Test: make, HFP regression test, testplans/79248
    Change-Id: I1a5eddceca0fc3e9e66e87337d7f7dd7415d6a87
    (cherry picked from commit fad35a48351d36224c7d4747cc5f1e7e1677d7fd)

**Comment by jerpelea** [comment]
the bug is fixed in android 9 and the patches can be backported locally

**Comment by dimitar10** [comment]
@richardtullberg thanks for the explanation.
